llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,185 @@
+//===-- LoopIterCountAssumptions.cpp - add Loop assumptions------*- C++ -*-===//
+//
+// This file is licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// (c) Copyright 2024 Advanced Micro Devices, Inc. or its affiliates
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumption"
+
+using namespace llvm;
+namespace {
+
+/// get AddRec scalar evolution, if it can be found in Op
+const SCEVAddRecExpr *getAddRecSCEV(Value *Op, ScalarEvolution *SE) {
+  if (!SE->isSCEVable(Op->getType()))
+    return nullptr;
+
+  return dyn_cast<SCEVAddRecExpr>(SE->getSCEV(Op));
+}
+
+/// get the SCEV of the loop IV
+std::optional<std::pair<const SCEVAddRecExpr *, int>>
+getLoopCmpIVAddRecExpr(ICmpInst &LoopCmpInstr, ScalarEvolution *SE) {
+
+  const SCEVAddRecExpr *S0 = getAddRecSCEV(LoopCmpInstr.getOperand(0), SE);
+  const SCEVAddRecExpr *S1 = getAddRecSCEV(LoopCmpInstr.getOperand(1), SE);
+
+  if (S0 && S1) {
+    LLVM_DEBUG(
+        dbgs() << "LoopIterCountAssumptions-Warning: Encountered two ADDRec "
+                  "expressions in the Loop Compare instruction. Currently it "
+                  " is expected that the compare consists of the IV and a "
+                  "static bound, that is loop invariant.");
+    return std::nullopt;
+  }
+
+  if (S0)
+    return std::optional(std::make_pair(S0, 0));
+
+  if (S1)
+    return std::optional(std::make_pair(S1, 1));
+
```

Comment:
small nit: Can we have a debug message stating that we don't have scevs? Because we know when we have two....

Reply:
I moved all the help messages into the corresponding methods

Reply:
so now the run function is cleaned up